src/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -271,9 +271,7 @@ private static unsafe void CopyImpl(Array sourceArray, int sourceIndex, Array de
             if (destinationArray == null)
                 throw new ArgumentNullException("destinationArray");
 
-            int sourceRank = sourceArray.Rank;
-            int destinationRank = destinationArray.Rank;
-            if (sourceRank != destinationRank)
+            if (sourceArray.EETypePtr != destinationArray.EETypePtr && sourceArray.Rank != destinationArray.Rank)
```

Comment:
`EETypePtr` has a custom `operator==` that does non-trivial work - it is different from CoreCLR. It does not look like an improvement.